File: Programmers/2/135807. 숫자 카드 나누기/숫자 카드 나누기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

bool isValid(int value, const vector<int>& target)
{
    for (int num : target)
        if (num % value == 0)
            return false;
    
    return true;
}

int gcd(int a, int b)
{
    while (b != 0)
    {
        int tmp = a % b;
        a = b;
        b = tmp;
    }
    
    return a;
}

int calculate(const vector<int>& arr)
{
    int result = arr[0];
    
    for (int i = 1; i < arr.size(); i++)
        result = gcd(result, arr[i]);
    
    return result;
}

int solution(vector<int> arrayA, vector<int> arrayB) {
    int answer = 0;
    
    int gcdA = calculate(arrayA);
    int gcdB = calculate(arrayB);
    
    if (isValid(gcdA, arrayB))
        answer = gcdA;
    
    if (isValid(gcdB, arrayA))
        answer = max(answer, gcdB);
    
    return answer;
}

File: Programmers/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
import java.util.*;

class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int answer = 0;
        int gcdA = calculateGCD(arrayA);
        int gcdB = calculateGCD(arrayB);
        
        if (isValid(gcdA, arrayB)) {
            answer = gcdA;
        }
        
        if (isValid(gcdB, arrayA)) {
            answer = Math.max(answer, gcdB);
        }
        
        return answer;
    }
    
    private int calculateGCD(int[] array) {
        int result = array[0];
        
        for (int i = 1; i < array.length; i++) {
            result = gcd(result, array[i]);
        }
        
        return result;
    }
    
    private int gcd(int a, int b) {
        while (b != 0) {
            int tmp = a % b;
            a = b;
            b = tmp;
        }
        
        return a;
    }
    
    private boolean isValid(int value, int[] targetArray) {
        for (int num : targetArray) {
            if (num % value == 0) {
                return false;
            }
        }
        
        return true;
    }
}